lib/curlable.js

```diff
@@ -5,182 +5,320 @@ const bodyParser = require('body-parser');
 
-const Curlable = function(cmdline, options, optionalLogger) {
-  this.options = options;
-  this.logger = optionalLogger || (() => {});
-  this.startTime = Date.now();
-  this.spawnedProcess = child_process.spawn('/bin/bash', ['-c', cmdline]);
+class Curlable {
+  constructor(cmdline, options) {
+    this.options = options;
+    this.logger = options && options.logger || (() => {});
+    this.terminationCallback = options && options.terminationCallback;
+    this.errorCallback = options && options.errorCallback;
+    this.prompt = options && options.prompt;
 
-  this.queriesQueue = null;
+    this.startTime = Date.now();
+    this.spawnedProcess = child_process.spawn('/bin/bash', [ '-c', cmdline ]);
 
-  this.nextLineHandler = null;
-  this.ready = true;
+    this.logger('Spawned `/bin/bash` with `' + cmdline + '`, PID ' + this.spawnedProcess.pid + '\n');
 
-  if (options && options.prompt) {
-    this.ready = false;
-    this.logger('Waiting for the external tool to be ready.\n');
-    this.nextLineHandler = (line) => {
-      if (line === options.prompt) {
-        this.ready = true;
-        this.logger('Ready after ' + (Date.now() - this.startTime) + 'ms.\n');
-        this.nextLineHandler = null;
+    this.queriesQueue = null;
+    this.currentQueueItem = null;
+
+    this.ready = true;
+
+    if (options && options.prompt) {
+      this.ready = false;
+      this.logger('Waiting for the external tool to be ready.\n');
+    }
+
+    this.spawnedProcess.on('close', (code) => {
+      if (!this.spawnedProcess) {
+        return;
       }
-    };
+
+      this.logger('Spawned process exited with code ' + code + '.\n');
+
+      this.spawnedProcess = null;
+      this._onSpawnedProcessEnd(new Error('The external tool exited with code ' + code + '.'));
+    });
+
+    readStreamByLines(
+      this.spawnedProcess.stdout,
+      (line) => {
+        this._onSpawnedProcessLine(line);
+      },
+      () => {
+        this.logger('Spawned process output stream has ended.\n');
+
+        this._onSpawnedProcessEnd(new Error('The external tool output stream has ended.'));
+      }
+    );
+  }
+
+  _finishQueueItem(queueItem, error) {
+    const resultTimestamp = Date.now();
+    const result = queueItem.accumulatedOutput;
+
+    // If the result is not empty, it has a '\n' already.
+    this.logger(queueItem.id + '\t' + (resultTimestamp - queueItem.writeTimestamp) + 'ms\t' +
+      (error ? 'ERROR: ' + error.message + '\n' : (result || '\n'))
+    );
+
+    const callback = queueItem.callback;
+    callback(null, result);
+
+    this._processQueue();
   }
 
-  readStreamByLines(this.spawnedProcess.stdout, (line) => {
-    if (!this.nextLineHandler) {
-      process.stderr.write('FATAL: No `nextLineHandler`, the tool' +
-        ' is generating output when it is not expected.\n');
-      process.stderr.write('ERROR: ' + line + '\n');
-      process.exit(-1);
-    } else if (line) {
-      this.nextLineHandler(line);
+  _onSpawnedProcessEnd(reason) {
```

Comment:
This name is misleasing, at least to me. "On" something reads as a callback (ex. "OnNewMessage()", "OnClick()", etc.), whereas here this method is something you are calling explicitly. Intentional?

Reply:
Intentional because this is kinda event handler, it's just the code you can see notifies about it.

It could have an imperative name e.g. `_handleSpawnedProcessEnd`, interchangeably. It's a matter of taste at this point.